### PR TITLE
Adapt dart2wasm

### PR DESCRIPTION
**Pull request by huanghui1998hhh**
### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [x] I have updated the documentation (if necessary)
- [x] I have run the tests without failures
- [x] I have updated the `CHANGELOG.md` in the corresponding package

### CODE REVIEW

File: dio/pubspec.yaml [hunk 1]
```diff
@@ -24,6 +24,7 @@ dependencies:
   http_parser: ^4.0.0
   meta: ^1.5.0
   path: ^1.8.0
+  web: ^0.5.1
```
Comment (kuhnroyal) [inline dio/pubspec.yaml:27]: What is the min SDK requirement to use the web package?

File: dio/lib/src/adapters/browser_adapter.dart [hunk 2]
```diff
@@ -135,16 +140,18 @@ class BrowserHttpClientAdapter implements HttpClientAdapter {
...
 
       final onSendProgress = options.onSendProgress;
       if (onSendProgress != null) {
-        xhr.upload.onProgress.listen((event) {
-          if (event.loaded != null && event.total != null) {
-            onSendProgress(event.loaded!, event.total!);
-          }
-        });
+        void onSendProgressWrap(web.ProgressEvent event) {
+          onSendProgress(event.loaded, event.total);
```
Comment (kuhnroyal) [inline dio/lib/src/adapters/browser_adapter.dart:151]: The progress loaded and total events are now non-nullable?

File: dio/lib/src/adapters/browser_adapter.dart [hunk 3]
```diff
@@ -111,15 +114,17 @@ class BrowserHttpClientAdapter implements HttpClientAdapter {
...
     if (requestStream != null) {
       if (connectTimeoutTimer != null) {
-        xhr.upload.onProgress.listen((event) {
+        void cancelConnectTimeoutTimer(web.ProgressEvent _) {
           connectTimeoutTimer?.cancel();
           connectTimeoutTimer = null;
-        });
+        }
+
+        xhr.upload.onprogress = cancelConnectTimeoutTimer.toJS;
```
Comment (kuhnroyal) [inline dio/lib/src/adapters/browser_adapter.dart:122]: Is there a way to pass these functions anonymously?

**Review by kuhnroyal — COMMENTED** [review]
Thanks!
Left some questions, I am generally not sure about min SDK support for this and if the tests work. We might need some test for the new header function.

**Comment by kuhnroyal** [comment]
Looks like we have to move to Dart 3.1.0 for this.

**Comment by AlexV525** [comment]
What about we split the Web support as a package and publish it first, then make a major version upgrade for wasm?

File: dio/pubspec.yaml (hunk 1 above)
Comment (kuhnroyal) [reply dio/pubspec.yaml:27]: Seems to be 3.1.0

**Comment by kuhnroyal** [comment]
Could work but that would also imply a breaking change for the main package.

**Comment by AlexV525** [comment]
> Could work but that would also imply a breaking change for the main package.

As I imagine it won't, we can set the constraint to allow both previous and the new version

File: dio/lib/src/adapters/browser_adapter.dart (hunk 3 above)
Comment (huanghui1998hhh) [reply dio/lib/src/adapters/browser_adapter.dart:122]: Yes, we can write like `xhr.upload.onprogress = (web.ProgressEvent event){}.toJS` this.

File: dio/lib/src/adapters/browser_adapter.dart (hunk 2 above)
Comment (huanghui1998hhh) [reply dio/lib/src/adapters/browser_adapter.dart:151]: > The progress loaded and total events are now non-nullable?

Yes, they are non-nullable in `package:web`.

File: dio/pubspec.yaml (hunk 1 above)
Comment (huanghui1998hhh) [reply dio/pubspec.yaml:27]: > Seems to be 3.1.0

Maybe 3.3.0?

**Comment by kuhnroyal** [comment]
We have to re-export some of the web classes in dio.

**Comment by AlexV525** [comment]
Offline discussed with @huanghui1998hhh , he will go ahead to try split the package. As I & he has tested the current package does not affect the `flutter build`, only affects the runtime.

File: dio/lib/src/adapter.dart [hunk 4]
```diff
@@ -4,7 +4,8 @@ import 'dart:typed_data';
 import 'package:meta/meta.dart';
 
 import 'adapters/io_adapter.dart'
-    if (dart.library.html) 'adapters/browser_adapter.dart' as adapter;
+    if (dart.library.js_util) 'adapters/web_adapters/web_adapter.dart'
```
Comment (Rexios80) [inline dio/lib/src/adapter.dart:7]: This should be `dart.library.js_interop` as mentioned here: https://dart.dev/interop/js-interop/package-web#conditional-imports

and everywhere else in the dio package

**Comment by AlexV525** [comment]
Closing in favor of #2223

**Pull request closed without merging**